datastore/adapter/keytransform.py

```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
+
+	Incoming keys are sharded into namespaces according to specifications
+	in a `SHARDING` file, which `prefix`, `suffix`, and `next-to-last`
+	sharding functions available.
+
+	The file should be be a single line in the following format:
+		{prefix}/version/function_name/key_length
+	
+	Implementation Notes:
+		* The `default_prefix` is "/repo/flatfs/shard/"
+		* The only version supported is "v1"
+		* function_name must be `prefix`, `suffix`, or `next-to-last`
+
+	For example:
+
+		>>> import datastore
+		>>>
+		>>> ds = datastore.DictDatastore()
+		>>> np = datastore.FlatFSDatastore(ds)
+		>>>
+		>>> np.put(datastore.Key('/abcdefghijk'), 1)
+		>>> np.get(datastore.Key('/abcdefghijk'))
+		1
+		>>> ds.get(datastore.Key('/abcdefghijk'))
+		None
+		>>> ds.get(datastore.Key('/ab/cd/ef/abcdefghijk'))
+		1
+		>>> np.put(datastore.Key('abc'), 2)
+		>>> np.get(datastore.Key('abc'))
+		2
+		>>> ds.get(datastore.Key('/ab/ca/bc/abc'))
+		2
+	"""
+
+	_default_sharding_file: str = "SHARDING"
+	_default_prefix: str = "/repo/flatfs/shard/"
+	
+	@staticmethod
+	def _prefix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).ljust(length, "_")[:length])
+	
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	def __init__(self, *args,
+	             sharding_file: str = None,
+	             prefix: str = None,
+	             **kwargs):
+		"""Initializes _FlatFSAdapter with sharding function.
+
+		Arguments
+		---------
+		sharding_file:
+			Path to the sharding file
+		prefix:
+			Prefix used in sharding file
+		"""
+
+		# assign the nesting variables
+		self.sharding_file = sharding_file if sharding_file is not None else self._default_sharding_file
+		self.prefix = prefix if prefix is not None else self._default_prefix
+
+		_sharding_function = self._parse_sharding_function()
+		super().__init__(*args, key_transform=_sharding_function, **kwargs)
+
+
+	def _parse_sharding_function(self) -> KEY_TRANSFORM_T:
+		"""Determine the sharding function from the SHARDING file.
+		
+		Raises
+		------
+		Exception
+			Sharding file was not found
+		Exception
+			Empty shard identifier file.
+		Exception
+			Prefix was not present in sharding file
+		Exception
+			Sharding file was in incorrect format
+		Exception
+			Expect `v1` format
+		Exceptiono
+		RuntimeError
+			An internal error occurred"""
+		# Ensure existence of sharding file
+		path: Path = Path(self.sharding_file)
+		if not path.exists():
+		    raise Exception(f"{self.sharding_file} file does not exist")
+
+		with open(path, "r") as f:
+		    sharding_func: str = f.read()
+
+		# Ensure not empty
+		if not sharding_func:
+		    raise Exception("Empty shard identifier file.")
+
+		sharding_func = sharding_func.strip()
```

Comment:
You need to read this from the backend datastore using something like

```py
sharding_func: str
try:
    # Try reading existing sharding func used
    sharding_func = self.child_datastore.get_all(self.sharding_key)
except KeyError:
    # Use the given default sharding func, remembering it for subsequent use of this repo
    sharding_func = self.default_sharding_func
    await self.child_datastore.put(self.sharding_key, sharding_func)
```

Where `self.default_sharding_func` is a parameter given in `__init__`. (Possibly with a default value.)

Reply:
Using `put` will require that we `await`, making the `_parse_sharding_func` method `async`, requiring that we `await` in `__init__`, making `__init__` `async`, which will throw:
```
TypeError: __init__() should return None, not 'coroutine'
```

Reply:
I think I have a gap in my understanding: To have access to `self. child_datastore` we need to inherit from `Datastore.ADAPTER_T`, where it gets [initialized](https://github.com/ipfs/py-datastore/blob/e566d40a8ca81d8628147e255fe7830b5f928a43/datastore/core/objectstore.py#L348), but `_Adapter` currently inherits from `typing.Generic[DS]`.

Reply:
`_Adapter` is not used by itself, but always a mixin class in `BinaryAdapter` and `ObjectAdapter`. These two types are conceptually very similar, but still strictly distinct types (they “look” very similar in terms of API, but represent different things). Since object and binary adapters are so similar API-wise when can actually implement both of these using just one piece of code, however the *meaning* of this code will vary depending on which context it is used it.

**Maybe this motivational story makes it clearer**: You, as a developer, want to expose a new datastore adapter (which you are doing) and to do that you'll need to implement two similar interfaces: `ObjectAdapter` and `BinaryAdapter`. They are very similar but not quite the same. So similar in fact, that if you were them implement them separately anyway, you would end up with what is essentially code duplication. You could move the identical code to separate functions, but now all your methods are left to do is calling those functions, and again the code that does this looks identical in both implementations. So you think: “Maybe I can somehow implement *both* of these interfaces in the *same* class?” So you create a new class and move all code there, but you cannot decide whether this should now be an `ObjectAdapter` or a `BinaryAdapter` – It's kinda both, or rather it *could* be both, but it's not both *at the same time*. You'll still need expose two separate classes – one for each type – they should just share the same method implementations. Turns out Python has a mechanism for exactly that: Multiple Inheritance! By inheriting from *both* the base class you'd like to implement and the extra class containing the just the actual method code, you end up with sharing the *code* between the two classes while still keeping them separate from a type perspective.

In other words: Think about the `_Adapter` class of just being a *container* for storing the methods you'd like to be visible in both the *ObjectAdapter* and the *BinaryAdapter* implementation.

And ignore that we are inheriting from `typing.Generic[DS]` there, that's just something make the types used during type checking (`tox -e typeck`) more accurate.

Reply:
Thank you for a wonderful explanation of Mixins. Makes a ton of sense.